### PR TITLE
fix: do not pass error to stream.destroy in sink

### PR DESCRIPTION
**Pull request by achingbrain**
Don't pass the error to the stream's destroy method as that would cause an error event to be emitted by the stream on the next tick by which point we've removed our error listener and it can result in an uncaught error.

### CODE REVIEW

**Comment by achingbrain** [comment]
Fixes https://github.com/ipfs/js-ipfs/issues/3169

cc: @jacobheun @gozala

File: sink.js [hunk 1]
```diff
@@ -72,19 +72,21 @@ module.exports = writable => async source => {
...
       writable.end()
     }
 
     // Wait until we close or finish. This supports halfClosed streams
     await waitForDone()
+  } catch (err) {
+    // the source threw, destroy the stream and throw the error
+    // call end too as we want it to be marked `writable: false`
+    writable.end()
+    writable.destroy()
```
Comment (achingbrain) [inline sink.js]: PassThrough streams (at least) do not set `.writable` to false when destroyed which seems quite naughty as the node docs say about `.writable`:

> Is true if it is safe to call `writable.write()`, which means the stream has not been destroyed, errored or ended

yet also about `.destroy([error])`:

> After this call, the writable stream has ended and subsequent calls to write() or end() will result in an ERR_STREAM_DESTROYED error.

So you'd think `.writable` would be false after `.destroy([error])` to avoid a `ERR_STREAM_DESTROYED` error but it is not:

```javascript
const { PassThrough } = require('stream')

const p = new PassThrough()

console.info(p.writable) // true
console.info(p.destroyed) // false

p.destroy()

console.info(p.writable) // true
console.info(p.destroyed) // true

p.end()

console.info(p.writable) // false
console.info(p.destroyed) // true
```

File: test/sink.test.js [hunk 2]
```diff
@@ -149,5 +149,6 @@ test('should destroy writable stream if source throws', async t => {
   const err = await t.throwsAsync(pipe(source, toIterable.sink(stream)))
 
   t.is(err.message, 'boom')
+  t.false(stream.writable)
```
Comment (jacobheun) [inline test/sink.test.js:315]: This test passes on master

File: sink.js (hunk 1 above)
Comment (jacobheun) [reply sink.js]: The reason there are separate try catch statements here is because a signular try/catch does not guarantee the source was the thing that errored. It's possible `waitForDone` can throw from the writable. Leaving this as 2 statements and changing the first catch to:
```
    writable.end()
    writable.destroy()
    error = err
```
would replicate what you're trying to do. However, it would be ideal to have a test that replicates the actual problem to verify this is working.

Comment (achingbrain) [reply sink.js]: Looks like `waitForDrainOrClose` can also cause a throw from an error on the writable, and that's caught by the try/catch as if it were from the source too.

Comment (achingbrain) [reply sink.js]: I've updated this to restore the second try/catch

File: test/sink.test.js (hunk 2 above)
Comment (achingbrain) [reply test/sink.test.js:315]: I've added a bunch of assertions to make it clear when we expect the stream to be made non-writable and when we expect it to be destroyed, so some of those will pass already.

I've also added more tests that try to hit some of the edge cases and also added tests for slow iterators because we were finishing our `source` in the tests before error events were being emitted that otherwise lead to crashes.

Comment (achingbrain) [reply test/sink.test.js:315]: > I've added a bunch of assertions to make it clear when we expect the stream to be made non-writable and when we expect it to be destroyed

You know, I say that, and then node 14 goes and changes `.destroyed` to true all on its own when the `close` event is emitted.  Uff.

Comment (achingbrain) [reply test/sink.test.js:315]: Turns out node 14 flips the `autoDestroy` option default value to true.  I've specified it as false in the tests so we can assert on whether we've destroyed the stream or not.

**Review by alanshaw — APPROVED** [review]
👌
